### PR TITLE
Bug fix for Java Functional Test - Using timestamp to retry

### PR DESCRIPTION
**Pull request by issacnitin**

### CODE REVIEW

File: java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java [hunk 1]
```diff
@@ -32,8 +33,11 @@ public void testAssertTitleWithSelenium() throws InterruptedException, Assertion
             System.out.println(isAlertPresent());
             driver.switchTo().alert().accept();
         }
-        int numRetries = 5;
-        for (int i = 0; i < numRetries; i++)
+
+        long currentTimestamp = (new Date()).getTime();
+        long endTimestamp = currentTimestamp + 60*10*1000;
+
+        while(currentTimestamp <= endTimestamp + 10000)
```
Comment (sachinma) [inline java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java]: why are you adding 10000 again

File: java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java [hunk 2]
```diff
@@ -43,12 +47,11 @@ public void testAssertTitleWithSelenium() throws InterruptedException, Assertion
             }
             catch(AssertionError e)
             {
-                if(i == (numRetries - 1))
-                {
+                if(currentTimestamp > endTimestamp) {
```
Comment (sachinma) [inline java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java]: why are you using the old value of currentTimestamp here

File: java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java (hunk 1 above)
Comment (issacnitin) [reply java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java]: 10000 milliseconds as buffer

File: java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java (hunk 2 above)
Comment (issacnitin) [reply java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java]: So exception should be thrown if test runs for more than 10 minutes, endTimestamp is calculated as initial currentTimestamp + 10 minutes. With 10 seconds as buffer, if currentTimestamp is > endTimestamp, we throw error

File: java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java [hunk 3]
```diff
@@ -32,8 +33,11 @@ public void testAssertTitleWithSelenium() throws InterruptedException, Assertion
             System.out.println(isAlertPresent());
             driver.switchTo().alert().accept();
         }
-        int numRetries = 5;
-        for (int i = 0; i < numRetries; i++)
+
```
Comment (bishal-pdMSFT) [inline java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java]: Please increase page load timeout to 360 seconds in line no 27

File: java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java (hunk 1 above)
Comment (bishal-pdMSFT) [inline java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java]: no need to have a condition here. You an have while(1). Instead use the condition in catch block

File: java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java (hunk 2 above)
Comment (bishal-pdMSFT) [inline java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java]: Wouldn't it be better to use the above condition here. You can calculate currentTimestamp here

File: java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java [hunk 4]
```diff
@@ -43,12 +47,11 @@ public void testAssertTitleWithSelenium() throws InterruptedException, Assertion
             }
             catch(AssertionError e)
             {
-                if(i == (numRetries - 1))
-                {
+                if(currentTimestamp > endTimestamp) {
                     throw e;
                 }
-                Thread.sleep(5000);
```
Comment (bishal-pdMSFT) [inline java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java]: Don't remove sleep

File: java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java (hunk 1 above)
Comment (issacnitin) [reply java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java]: I'm not sure what you meant, If we use the condition in catch block to break out of the loop, in the case where webapp is successfully deployed, the loop will run forever.

**Comment by issacnitin** [comment]
I've updated the logic @sachinma @bishal-pdMSFT

File: java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java [hunk 5]
```diff
@@ -32,8 +33,11 @@ public void testAssertTitleWithSelenium() throws InterruptedException, Assertion
             System.out.println(isAlertPresent());
             driver.switchTo().alert().accept();
         }
-        int numRetries = 5;
-        for (int i = 0; i < numRetries; i++)
+
+        long currentTimestamp = (new Date()).getTime();
```
Comment (bishal-pdMSFT) [inline java/spring/containerWithTests/Application/sampleWebApp.FunctionalTests/src/main/java/functional_tests/SampleFunctionalTest.java]: call it startTimestamp

**Pull request merged**